Calling

ClaireGuerin opened the pull request:
@thejasvibr 

Hey Thejasvi, 

So this is the latest version, where the hearing part does not work... Feel free to add comments, so that I can modify it until we can accept the pull request of the code into the master branch!

ClaireGuerin committed:
testing for call

ClaireGuerin committed:
updated initializing param

ClaireGuerin committed:
callstarttime can be =! 0

ClaireGuerin committed:
correct a few coding mistakes

ClaireGuerin committed:
Add GitHub hand-out - .docx version

ClaireGuerin committed:
replaced abscissae and ordinate with x and y

ClaireGuerin committed:
set a proper environment launcher

ClaireGuerin committed:
add a for loop for multiple agents, but don't know how to store the results

ClaireGuerin committed:
Store instances in dictionary
the only problem is that I get the following error:

  File "C:/Users/Claire/Documents/GitHub/Bat-Project/bat_class.py", line 151, in __init__
    self.initpos = env.all_initpos[self.ID]

TypeError: list indices must be integers, not numpy.float64

ClaireGuerin committed:
Problem Solved!

ClaireGuerin committed:
Time resolution + movement for each bat instance

ClaireGuerin committed:
General plotting

ClaireGuerin committed:
Updated descriptions of functions and classes

ClaireGuerin committed:
Cleaned-up a bit

ClaireGuerin committed:
Adapted code so that it is compatible with TOA function

ClaireGuerin committed:
No crucial changes

ClaireGuerin committed:
removed some append

ClaireGuerin committed:
started hearing part

ClaireGuerin committed:
Circle thickness

ClaireGuerin committed:
Added a limitation in storing time of calls in TOA function

ClaireGuerin committed:
update radius of each call according to the actual time

ClaireGuerin committed:
Created specific update function for our dict & started turning TOA into a class

ClaireGuerin committed:
Minor changes

ClaireGuerin committed:
Turned TOA into class

ClaireGuerin committed:
Reduced TOA function into smaller separate functions within Hearing_00 class

ClaireGuerin committed:
deleted docx file

ClaireGuerin committed:
Added detailed description of the class Hearing_00

ClaireGuerin committed:
not fully functional yet

ClaireGuerin committed:
Solve: Data_storage. New Pb: Propagation

ClaireGuerin committed:
Renaming init output and make sure floats and integers are of the good format

ClaireGuerin committed:
Nested classes Launcher and Jamming

ClaireGuerin committed:
doc/program

ClaireGuerin committed:
Merge branch 'master' into calling

ClaireGuerin committed:
Changed max and min circle functions for a global ring test

ClaireGuerin committed:
Nested classes, time resolution fixed

ClaireGuerin committed:
Description of Launcher class

ClaireGuerin committed:
Updated the description of all functions and classes

ClaireGuerin committed:
Erased unnecessary comments

ClaireGuerin committed:
Key error when running the code: problem in Propagation function...?

ClaireGuerin committed:
Minor changes

ClaireGuerin committed:
Direct comments on code updated

ClaireGuerin committed:
Renamed file

ClaireGuerin committed:
Adapted calling function to time resolution

ClaireGuerin committed:
Working version... Except for hearing!

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+    def __init__(self, popsize, boxsize, d_t, simduration):
+        self.popsize = int(popsize)
+            # integer. Size of the bat population / Number of agents 
+        self.d_t = float(d_t * m.pow(10,-3))
```
1) pls change variable name to sthing meaning - 'timeres' or so ? 
2) keep the time scale flexible - and let it be in seconds - so if we want to deal with ms then we just x 10^-3, or if it's in microseconds , then it's 10^-6 etc etc

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+        
+        assert isinstance(self.boxsize, list) and len(self.boxsize) == 2, "'boxsize' must be a list of 2 elements." 
+            # returns error if boxsize is not of the right format, i.e. [x,y]
+        
```
pls add comment: Identification: a function which assigns a unique serial number to each agent in the population

ClaireGuerin replied:
@thejasvibr You mean for Identification only, or for every function in the code? Because there is already a detailed description of the functions and classes in the README file. I thought it would make the code lighter, and if you need the detailed description you still have access to it.

thejasvibr replied:
I'd still add in a single but brief description of what it's doing. It's nice to be able to read a quick summary right next to the code.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+            self.all_ID[int(agent)] = agent
+                # fill-in a list of all agents ID 
+                # from 0 to n = (population size - 1)
+            
```
change the random position placement - maybe assign the positions of the agents based on a matrix or list of positions that you give in when you initiate the simulation.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+           self.all_initpos[int(ID),0] = self.agent_initx
+               # store the x-coordinate in an array
+           self.all_initpos[int(ID),1] = self.agent_inity
+               # store the y-coordinate in an array
```
change the random position placement - maybe assign the positions of the agents based on a matrix or list of positions that you give in when you initiate the simulation.

ClaireGuerin replied:
@thejasvibr But the thing is, each time I change the number of agents, I have to change the matrix too. I don't think it's very efficent time-wise, and the calculations of distance that you've made earlier can be included somewhere really easily. I don't really feel like it would change that much the analysis or expectations, once we know the inter-individual distances

thejasvibr replied:
Agreed- if you have a small piece of code that makes an inter-individual distance matrix , and then calculates the time delay from that (just divide the matrix by speed of sound) - you'd be able to get the same info - without having to start out with defined positions. Either way - it's important to have this data to get a handle of how/whether the function is working or not.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # store the new, incremented time count in timeclock.
+
+    class Bat_Jamming_00:
+        # Class implementation nested into Launcher class implementation.
```
this is pretty cool. Does this mean you can have multiple instances of 'Launcher' running on your computer (with each instance having their own instance of the 'Bat_Jamming_00') ?

ClaireGuerin replied:
Yes! Haven't tried it yet, but absolutely :) Although in Bat_Jamming_00 I'll have to change the use of variables from Launcher , cause we can't run several simulations at the same time if they're all called the same ("env")

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # Must be between -pi and pi (asserted later).
+            self.stepsize = float(flightspeed * self.d_t) 
+                # float. Distance in meters covered by the agent in 1 time step
+            self.IPI = int(float(IPI * m.pow(10,-3)) / env.d_t)
```
here too - it's good to keep code general - keep things in terms of seconds, instead of milliseconds - this allows you to change parameters comfortably to any range you want without causing confusion.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # float. Time maximal, in seconds, during which a sound can 
+                # travel, before its intensity passes below the hearing 
+                # threshold of the agent.
+            self.ring_width = float(1) 
```
it's fine for now. In later versions this should depend on the call duration: ring_width= callduration \* velocity of sound

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # travel, before its intensity passes below the hearing 
+                # threshold of the agent.
+            self.ring_width = float(1) 
+                # float. Spatial difference between start and end of call (m).
```
'spatial difference' feels vague - perhaps we want to say - the ring_width is the difference in radii of the two concentric circles ( in 2D) which form the start and the end of the bat call.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+            self.ring_width = float(1) 
+                # float. Spatial difference between start and end of call (m).
+            self.hearhistory = []
+                # empty array, dimensions 3*unknown. Will store the 
```
why are the dimensions 'unknown'.....maybe want to change it to say something like : 'dimensions 3*Number of sounds heard ?' - also perhaps you could make the decription more explicit by saying - 

"Row #1 stores the ID of the source agent's call, Row _2 stores the time at which the call was emitted Row_3: the time at which it was heard by the focal agent"

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+            self.calltest = float(self.timestep-self.firstcall)/float(self.IPI)
+                # calculate the theoretical number of calls since the 1st call:
+                # time since 1st call = current time - time of 1st call/IPI
+            
```
this does make sense mathematically - and is a neat implementation. From an 'biologically intuitive' perspective - could you change the way the self.calltest works - so that it keeps track of only the last call?  eg. 

self.lastcalltime= _code which looks at the last call by focal agent_
elapsedtime=timestep-lastcalltime 

if elapsedtime=IPI: initiate call
else : don't do anything. 
- and also an assert to check if the elapsed time if more than IPI :P

ClaireGuerin replied:
@thejasvibr I agree to the fact that this bit of code isn't intuitive biologically speaking, but I think that it's the straightest and simplest way to look at it. Although the example code that you've written in comment looks simple on first sight, "code which looks at the last call by focal agent" is very vague, and when digging-in a little bit, you'll realize how messy it can get computationally speaking. Honestly, I'm not even sure how to do this, unless I create a new for loop within the function... The question is, is it worth it to loose computational efficiency in order to make the code "biologically intuitive", although we have a straight way to obtain the same result? I'll make another file where I try to do it the "biological" way, just to see.

thejasvibr replied:
Agreed. For now - this could remain - the pre-calculation approach is fine - maybe you don't want to bother trying out another version right now. 
For later versions with feedback loops etc - this pre-calculation will not work - because the bat may need to react to local cues and change IPI.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+            if self.calltest%1 == 0:
+                # if the theoretical number of calls since the 1st call is a 
+                # natural number:
+                self.callshistory[int(self.timestep)] = 1
```
- I'm just wondering if you need to keep restating that some numbers or values are 'integers' eg. in the 
  int(self.timestep) -- it feels a bit repetitive - since you have already reclared the timestep to be a sequence of whole numbers.

ClaireGuerin replied:
@thejasvibr Probably not, and you're right to point it out. I just wanted to make sure there won't be any "translation" problem, as I'm using these parameters to index back in lists / dictionaries. I'll see where it is more sensible to specify that these are integers.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # update the propagation distance in dict1
+        
+            return dict1
+    
```
In_ring: function which tests if a bat is within the 'ring of sound' of a call

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # (radius - width)?
+        
+        def Hearing_test(self, ag_id, tcall, hear_array):
+            self.ringtest = self.In_ring(env.callsources[int(ag_id)][int(tcall)]['xsource'],
```
this is a bit scary to read. Could you re-write the code so its semantics is clearer. 
eg:
agent_xpos=int(ag_id)][int(tcall)]['xsource'] --- here too -  am wondering if there's  a need to keep restating 'int' on the self.ID number - even though it is already an integer from the start by definition ? 
agent_ypos=int(ag_id)][int(tcall)]['ysource']
etc
etc
self.ringtest= self.In_ring(agent_xpos,agent_ypos,callcentre_x,callcentre_y)

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+        
+            return dict1
+    
+        def In_ring(self, center_x, center_y, radius, x, y):
```
could you change the name of the center_x to reflect the fact that you're referring to the call: eg. callcentre_x & callcentre_y - and also change the 'x' and 'y' to - agent_x and agent_y

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+            return dict1
+    
+        def In_ring(self, center_x, center_y, radius, x, y):
+            dist = (x - center_x) ** 2 + (y - center_y) ** 2
```
you're getting a square of the distance here (metre^2)- need to sqrt it to get distance in metres.

thejasvibr commented on README.txt:
```diff
@@ -1,207 +1,186 @@
+# Inputs: center_x, center_y, radius, x, y
+# center_x: float. Abscissa of the source of the call
+# center_y: float. Ordinate of the source of the call
+# radius: float. Distance of the propagated call to its source 
```
pls change the 'radius' to something like - distfrmsource - just 'radius' feels a bit too general.
maybe change the wording to : distance of the propagated call _from_ its source

thejasvibr commented on README.txt:
```diff
@@ -1,207 +1,186 @@
+#----------IN_RING----------#
+# Function that tests whether a certain point (agent) is on the ring of sound.
+# Inputs: center_x, center_y, radius, x, y
+# center_x: float. Abscissa of the source of the call
```
center_x is not so informative - could you change name to:
callsource_x and callsource_y

thejasvibr commented on README.txt:
```diff
@@ -1,207 +1,186 @@
+
+#----------IN_RING----------#
+# Function that tests whether a certain point (agent) is on the ring of sound.
+# Inputs: center_x, center_y, radius, x, y
```
pls change the 'radius' to something like - distfrmsource - just 'radius' feels a bit too general.

thejasvibr commented on sensory_jamming.py:
```diff
@@ -0,0 +1,342 @@
+                # distance between an agent and a call's source
+            return dist > radius and dist < radius - self.ring_width
+                # boolean. Is dist within the distance travelled by the call 
+                # between the beginning (radius) and the end of the call 
```
1) Perhaps put brackets to make the boolean tests clearer
2) I think it should be dist <radius (the bat should be 'behind' the start of the call) AND  [dist > (radius-ring_width)] - the bat should be 'ahead' of the end of the call

ClaireGuerin committed:
Changed In_ring function

ClaireGuerin committed:
@thejasvibr I made most of the changes you requested in your comments.
I didn't always agree with your comments, so I explained why I didn't make the corresponding changes in answers to your comments.
Of course we can discuss it further, my mind is not made yet on everything! ;)

About the Hearing Function: it still doesn't work (as in, it still returns an empty list for hearhistory)

 :wq

ClaireGuerin commented:
@thejasvibr I added some new commits with changes according to your comments (see message of last commit for more details). The Hearing part is still malfunctionning though, despite the changes in In_Ring function. I don't know a lot about debugging, so I'm reading through tutorials to see how to find the origin of the error (I created a file called Debugging in Python in Google Drive's Coding folder). I'll let you know if anything comes out!

ClaireGuerin committed:
This version works

ClaireGuerin committed:
Exporting data

ClaireGuerin committed:
New file: data_analysis

ClaireGuerin committed:
Import calls and movements data into R

ClaireGuerin committed:
Working on Hearing data import

ClaireGuerin committed:
Fixed an error in data import

ClaireGuerin committed:
Fixed another data exportation problem

ClaireGuerin committed:
Some changes

ClaireGuerin committed:
Ask user for coordinates input of each agent

ClaireGuerin committed:
Decomposed loop into separate smaller loops for calling hearing and movement respectively

ClaireGuerin committed:
Trying to plot ring

ClaireGuerin committed:
some changes

ClaireGuerin committed:
Merge branch 'calling' of https://github.com/ClaireGuerin/Bat-Project into calling

ClaireGuerin committed:
Plot bats' positions & rings of sound over time on a single plot

ClaireGuerin committed:
Bat's initial positions implemented as pre-defined matrix

ClaireGuerin committed:
"Rounding-off" issue that arose when registering the calls' durations in Hearing seems to be fixed.
What has been modified:
the ring of sound is made larger, as it integrates the space that is occupied by the sound
for a time range (t) such as: x <= t < x+1 where x is the current time step and x+1 the next time step.
Thus, at time step x, the sound "has occupied" space from the end of the call at time x to the start of the call
at time "right before" x+1.

ClaireGuerin committed:
Modif matrix positions input + started modelling echoes

ClaireGuerin committed:
Err: Building-up echo processing. No echo registered.

ClaireGuerin committed:
Now echoes are reflected, but need to test if and when they're heard

ClaireGuerin committed:
Err: Building-in echo paths. Script runs, echoes are generated, but sound directly 'reflects' on bat itself...'

ClaireGuerin committed:
Err: Building-in echo paths.
Calls are reflected, generating echoes.
Change: in Impact_test, Impact test not run if the identity of the focal agent is actually the identity of the call emittor.
Problem: now echoes, although they're generated, are not heard by the bats O.o

ClaireGuerin committed:
Err: Building-in echo paths. Propagation distance is the same for call & its echo! NB: still need to verify that echoes come at the proper times

ClaireGuerin committed:
Image format

ClaireGuerin committed:
Err: Building-in echo reflection. When adding the condition the reflection only acts on others and not the emittor of the call itself, no echoes are produced anymore, and a number unconsistent of calls is registered
